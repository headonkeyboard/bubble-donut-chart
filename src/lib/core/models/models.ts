export interface Point {
  x: number;
  y: number;
}

export interface RawData {
  id: string;
  label: string;
  weight: number;
  group: string;
}

export interface Section {
  id: string;
  ratio: number;
  bubbles: BubbleWithCoordsAndRadius[];
  startAngle: number;
  endAngle: number;
  gridLength: number;
  edges: GridEdges;
}

export interface GridEdges {
  top: Point[];
  right: Point[];
  bottom: Point[];
  left: Point[];
}

export type Coord = Point & { r: number };
export type BubbleWithCoordsAndRadius = RawData & Coord;
export type Grid = Array<number>;
