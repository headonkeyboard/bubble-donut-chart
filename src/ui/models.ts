export interface Section {
    bubbleCount: number;
    bubbleMinValue: number;
    bubbleMaxValue: number;
}
