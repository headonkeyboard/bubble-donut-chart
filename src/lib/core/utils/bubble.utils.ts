import {
  RawData,
  Coord,
  Grid,
  GridEdges,
  Point,
  MaxDistancePoint,
} from "../models";
import {
  gridPointPositionToPixels,
  getDistance,
  getMinDistance,
} from "./coords.utils";

/**
 * Return a bubble radius based on it's weight proportional to donutArea
 *
 * @param bubbleValue RawData coefficient weight
 * @param donutArea Donut surface
 * @param totalBubbleValues Sum of all bubbles coefficient weight
 */
const getBubbleRadius = (
  bubbleValue: number,
  donutArea: number,
  totalBubbleValues: number
): number => {
  const bubbleArea = (bubbleValue * donutArea) / totalBubbleValues;
  return Math.sqrt(bubbleArea / Math.PI) * 0.7;
};

/**
 * Generate an array representing a donut section grid
 * Each item index is a grid coordinate
 * Each item value is equal to the minimum distance with an obstacle (edge or any other bubble)
 *
 * @param edges An array containing edges positions
 * @param gridLength The grid size
 * @param startAngle The section start angle in rad
 * @param endAngle The section end angle in rad
 * @param innerRadius The donut inner radius in pixels
 * @param outerRadius The donut outer radius in pixels
 * @param gridPointPositionsCache
 * @param grid
 * @param bubbleCoords Positions of bubbles already in grid
 */
const generateGrid = (
  edges: GridEdges,
  gridLength: number,
  startAngle: number,
  endAngle: number,
  innerRadius: number,
  outerRadius: number,
  gridPointPositionsCache: Map<number, Point>,
  grid: Grid = [],
  bubbleCoords: Array<Coord> = []
): {
  grid: Grid;
  maxDistancePoint: MaxDistancePoint;
} => {
  let maxDistance = 0;
  let maxDistanceX = 0;
  let maxDistanceY = 0;

  for (let i = 0; i < Math.pow(gridLength, 2); i++) {
    let minDistance = null != grid[i] ? grid[i] : null;

    // skiping minDistance search when already low
    if (null != minDistance && minDistance <= 0) {
      continue;
    }

    const x = i % gridLength;
    const y = Math.floor(i / gridLength);

    let gridPointCoord!: Point;

    if (gridPointPositionsCache.has(i)) {
      gridPointCoord = gridPointPositionsCache.get(i) as Point;
    } else {
      gridPointCoord = gridPointPositionToPixels(
        x,
        y,
        startAngle,
        endAngle,
        gridLength,
        innerRadius,
        outerRadius
      );

      gridPointPositionsCache.set(i, gridPointCoord);
    }

    // compute edge distance from grid point only for an empty grid
    if (null === minDistance) {
      // reduce edges list to potential closest edges
      const closestEdges: Point[] = [];

      if (y > Math.round(gridLength / 2)) {
        closestEdges.push(...edges.right);
      } else if (y <= Math.round(gridLength / 2)) {
        closestEdges.push(...edges.left);
      }

      if (x > Math.round(gridLength / 2)) {
        closestEdges.push(...edges.top);
      } else if (x <= Math.round(gridLength / 2)) {
        closestEdges.push(...edges.bottom);
      }

      minDistance = getMinDistance(closestEdges, gridPointCoord);
    }

    if (bubbleCoords.length > 0) {
      const lastAddedBubble = bubbleCoords[bubbleCoords.length - 1];
      const bubbleCoordDistance =
        getDistance(gridPointCoord, lastAddedBubble) - lastAddedBubble.r;
      if (minDistance && minDistance > bubbleCoordDistance) {
        minDistance = bubbleCoordDistance;
      }
    }

    if (minDistance > maxDistance) {
      maxDistance = minDistance;
      maxDistanceX = x;
      maxDistanceY = y;
    }

    grid[i] = minDistance;
  }

  return {
    grid,
    maxDistancePoint: {
      maxDistance,
      x: maxDistanceX,
      y: maxDistanceY,
    },
  };
};

/**
 * Return bubble coordinates and radius in pixels
 *
 * @param maxDistance
 * @param maxDistanceCoords
 * @param value The bubble value (the higher its value, the higher its radius)
 * @param donutArea The computed area of the donut
 * @param totalBubbleValues The sum of all bubble value
 */
const getBubblePositionAndRadius = (
  maxDistance: number,
  maxDistanceCoords: Point,
  value: number,
  donutArea: number,
  totalBubbleValues: number
) => {
  let bubbleRadius = getBubbleRadius(value, donutArea, totalBubbleValues);

  if (maxDistance - bubbleRadius < 0) {
    bubbleRadius = maxDistance - 1;
  }

  const offset = (maxDistance - bubbleRadius) / 2;

  const offsetX = offset * (Math.random() < 0.5 ? -1 : 1);
  const offsetY = offset * (Math.random() < 0.5 ? -1 : 1);

  return {
    bubbleX: maxDistanceCoords.x + offsetX * Math.random(),
    bubbleY: maxDistanceCoords.y + offsetY * Math.random(),
    bubbleR: bubbleRadius >= 0 ? bubbleRadius : 1,
  };
};

export { generateGrid, getBubbleRadius, getBubblePositionAndRadius };
