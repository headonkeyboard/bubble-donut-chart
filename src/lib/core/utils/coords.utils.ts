import { Grid, Point } from '../models/models';

/**
 * Use Pythagorean theorem to return a distance (in pixels) between two point
 */
const getDistance = (
  coord1: Required<Point>,
  coord2: Required<Point>
): number => {
  return Math.sqrt(
    Math.pow(coord1.x - coord2.x, 2) + Math.pow(coord1.y - coord2.y, 2)
  );
};

/**
 * Returns the min distance between a list of points (edges Param) and point
 *
 * @param points Array of positions
 * @param point The starting position
 */
const getMinDistance = (points: Point[], point: Required<Point>) => {
  const edgesDistances = points.reduce((acc: Array<number>, edge) => {
    acc.push(getDistance(edge, point));

    return acc;
  }, []);

  return Math.min(...edgesDistances);
};

/**
 * Return a couple of the max distance found in grid with it's corresponding grid coordinates
 *
 * @param grid Donut section grid
 * @param startAngle Donut section start angle in Radian
 * @param endAngle Donut section end angle in Radian
 * @param gridLength Number of grid points in a row or in a column
 * @param innerRadius Donut inner radius
 * @param outerRadius Donut outer radius
 */
const getMaxDistance = (
  grid: Grid,
  startAngle: number,
  endAngle: number,
  gridLength: number,
  innerRadius: number,
  outerRadius: number
): { maxDistance: number; maxDistanceCoords: Point } => {
  let maxDistance = grid[0];
  let maxDistanceCoords = getGridPointPosition(
    0,
    0,
    startAngle,
    endAngle,
    gridLength,
    innerRadius,
    outerRadius
  );

  for (let i = 0; i < grid.length; i++) {
    const x = i % gridLength;
    const y = i / gridLength;

    if (grid[i] > maxDistance) {
      maxDistance = grid[i];
      maxDistanceCoords = getGridPointPosition(
        x,
        y,
        startAngle,
        endAngle,
        gridLength,
        innerRadius,
        outerRadius
      );
    }
  }

  return { maxDistance, maxDistanceCoords };
};

/**
 * Return a position in pixels from grid coordinates
 *
 * @param x coord in Donut section
 * @param y coord in Donut section
 * @param startAngle Donut section start angle in Radian
 * @param endAngle Donut section end angle in Radian
 * @param gridLength Number of grid points in a row or in a column
 * @param innerRadius Donut inner radius
 * @param outerRadius Donut outer radius
 */
const getGridPointPosition = (
  x: number,
  y: number,
  startAngle: number,
  endAngle: number,
  gridLength: number,
  innerRadius: number,
  outerRadius: number
): Point => {
  const sectionAngle = endAngle - startAngle;
  const gridPointAngle = startAngle + (sectionAngle * y) / gridLength;
  const gridPointDistanceWithCenter =
    innerRadius + ((outerRadius - innerRadius) * x) / gridLength;

  return {
    x: gridPointDistanceWithCenter * Math.cos(gridPointAngle),
    y: gridPointDistanceWithCenter * Math.sin(gridPointAngle),
  };
};

const getPointWithRatio = (point: Point, ratio: number) => {
  return {
    x: point.x * ratio,
    y: point.y * ratio,
  };
};

export {
  getDistance,
  getGridPointPosition,
  getMaxDistance,
  getMinDistance,
  getPointWithRatio,
};
