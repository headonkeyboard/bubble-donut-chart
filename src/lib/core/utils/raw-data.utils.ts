import { RawData } from "../models";

/**
 * Return RawData entries grouped by their group property
 *
 * @param rawData
 */
export const getRawDataByGroup = (
  rawData: RawData[]
): Map<string, RawData[]> => {
  const rawDataByGroup = new Map<string, RawData[]>();

  rawData.forEach((entry) => {
    if (rawDataByGroup.has(entry.group)) {
      (rawDataByGroup.get(entry.group) as RawData[]).push(entry);
    } else {
      rawDataByGroup.set(entry.group, [entry]);
    }
  });

  return rawDataByGroup;
};

/**
 * Return the sum of weight property of all rawData entries
 *
 * @param rawData
 */
export const getRawDataWeightSum = (rawData: RawData[]): number => {
  return rawData.reduce((acc, entry) => acc + entry.weight, 0);
};
