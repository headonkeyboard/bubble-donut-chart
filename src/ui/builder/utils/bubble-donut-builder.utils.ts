/**
 * Return a new group size equal to 100/groupCount (including the new one)
 * This size make the equal distribution with other groups easier when adding a new group
 *
 * @param groupCount number of existing groups before adding a new one
 */
export const newGroupSize = (groupCount: number): number | void => {
    if (groupCount < 1) {
        throw new Error("Group count must be greater than one");
    }

    return Math.floor(100 / (groupCount + 1));
};


/**
 * Return new groups size after group at updateIndex has been updated with a given updateSize
 * Group at updateIndex is updated with updateSize but keeps its value in range [1, 100 - groupCount - 1]
 * Other groups update their size to share remaining percentage equally
 *
 * @param groupsSize Every group sizes in percent
 * @param updateIndex The index of the group being updated
 * @param updateSize The new value of the group being updated
 */
export const updatedGroupsSize = (groupsSize: number[], updateIndex: number, updateSize: number) => {
    const updatedGroupNewSize = Math.max(
        Math.min(updateSize, 100 - groupsSize.length - 1),
        1
    );
    const otherGroupsSizeSum = Math.max(100 - updatedGroupNewSize, 1);
    const otherGroupsNewSize = Math.floor(otherGroupsSizeSum / (groupsSize.length - 1));

    return groupsSize.map(
        (groupSize, groupIndex) => {
            if (groupIndex !== updateIndex) {
                return otherGroupsNewSize;
            } else {
                return updatedGroupNewSize;
            }
        }
    );
};
