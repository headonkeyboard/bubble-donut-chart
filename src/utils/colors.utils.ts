import * as d3 from "d3";

export const groupColors = (groupCount: number) => {
  const groups = [];

  for (let i = 0; i < groupCount; i++) {
    groups.push(`group ${i + 1}`);
  }

  return d3.scaleOrdinal().domain(groups).range(d3.schemeSet2);
};
