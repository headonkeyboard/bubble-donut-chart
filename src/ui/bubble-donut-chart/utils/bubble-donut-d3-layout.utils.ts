import * as d3 from "d3";
import {
  BubbleWithCoordsAndRadius,
  RawData,
} from "../../../lib/core/models/models";
import { BubbleDonut } from "../../../lib/core/bubble-donut";
import { groupColors } from "../../../utils/colors.utils";
import { getRandomInt } from "../../../lib/core/random.utils";
import { DEFAULT_RADIUS, DEFAULT_VIEWPORT, DONUT_CENTER } from "../constants";

export type D3SVGSelection = d3.Selection<
  SVGSVGElement,
  BubbleWithCoordsAndRadius,
  null,
  undefined
>;

/**
 * Init a new BubbleDonut from data
 *
 * @param data
 */
const rawDataToBubbleDonut = (data: RawData[]): BubbleDonut => {
  return new BubbleDonut(DEFAULT_RADIUS * 0.2, DEFAULT_RADIUS, data);
};

/**
 * Returns a function which will scale an input according to range [0, width]
 *
 * @param width
 */
const viewportScaleRange = (width: number) => {
  return d3.scaleLinear().domain([0, DEFAULT_VIEWPORT]).range([0, width]);
};

/**
 * Generate bubbles from data and layout them to build a Bubble Donut
 *
 * @param selection
 * @param width
 * @param data
 */
export function layoutBubbles(
  selection: D3SVGSelection,
  width: number,
  data: RawData[]
) {
  const bubbleDonut = rawDataToBubbleDonut(data);
  const d3Bubbles = bubbleDonut.getBubbles();
  const groupsColors = groupColors(bubbleDonut.sections.size);
  const viewportScale = viewportScaleRange(width);
  const d3BubbleSelection = selection
    .selectAll<SVGCircleElement, BubbleWithCoordsAndRadius>("circle")
    .data(d3Bubbles, (d) => (d as BubbleWithCoordsAndRadius).id);

  d3BubbleSelection
    .enter()
    .append("circle")
    .attr("opacity", 0)
    // make new circles coming from a random position to give a nice effect
    .attr("cx", () => viewportScale(getRandomInt(0, DEFAULT_VIEWPORT)))
    .attr("cy", () => viewportScale(getRandomInt(0, DEFAULT_VIEWPORT)))
    .attr("r", 0)
    .attr("fill", (d) => groupsColors(d.group) as string)
    .merge(d3BubbleSelection)
    .transition()
    .duration(800)
    .attr("cx", (d) => viewportScale(d.x + DONUT_CENTER))
    .attr("cy", (d) => viewportScale(d.y + DONUT_CENTER))
    .attr("r", (d) => viewportScale(d.r))
    .attr("opacity", 1)
    .attr("fill", (d) => groupsColors(d.group) as string);

  // makes bubbles moving to random position before removing theme (looks like an explosion)
  d3BubbleSelection
    .exit()
    .transition()
    .duration(400)
    .attr("cx", () => getRandomInt(0, width))
    .attr("cy", () => getRandomInt(0, width))
    .attr("r", 0)
    .remove();
}

/**
 * Create svg root container
 *
 * @param divWrapper
 * @param size
 */
export const createD3SvgRoot = (divWrapper: HTMLDivElement, size: number) => {
  return d3
    .select<HTMLDivElement, BubbleWithCoordsAndRadius>(divWrapper)
    .append("svg")
    .attr("height", size)
    .attr("width", size);
};

/**
 * Resize svg root container and all bubble children
 *
 * @param d3SvgRoot
 * @param size
 */
export const resizeD3ViewportAndBubbles = (
  d3SvgRoot: D3SVGSelection,
  size: number
) => {
  d3SvgRoot.attr("height", size).attr("width", size);

  const viewportScale = viewportScaleRange(size);

  d3SvgRoot
    .selectAll<SVGCircleElement, BubbleWithCoordsAndRadius>("circle")
    .attr("cx", (d) => viewportScale(d.x + DONUT_CENTER))
    .attr("cy", (d) => viewportScale(d.y + DONUT_CENTER))
    .attr("r", (d) => viewportScale(d.r));
};
