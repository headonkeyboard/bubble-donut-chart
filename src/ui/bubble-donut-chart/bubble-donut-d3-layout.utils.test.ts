import React from "react";
import {createD3SvgRoot, layoutBubbles} from "./bubble-donut-d3-layout.utils";
import {RawData} from "../../lib/core/models/models";

jest.useFakeTimers();

const data: RawData[] = [
    { id: '1', group: 'group 1', label: 'test', weight: 10 },
    { id: '2', group: 'group 1', label: 'test', weight: 10 },
    { id: '3', group: 'group 2', label: 'test', weight: 10 },
    { id: '4', group: 'group 2', label: 'test', weight: 10 },
];

test("createD3SvgRoot", () => {
    const wrapper = document.createElement('div');
    createD3SvgRoot(wrapper, 200);
    expect(wrapper).toMatchSnapshot();
});

test("layoutBubbles", () => {
    const wrapper = document.createElement('div');
    const selection = createD3SvgRoot(wrapper, 200);
    layoutBubbles(selection, 200, data);

    const selectionNodeChildren = selection.node().children;
    expect(selectionNodeChildren.length).toBe(data.length);

    for (let i = 0; i < selectionNodeChildren.length; i++) {
        const child = selectionNodeChildren.item(i);
        expect(child.tagName).toBe('circle');
        expect(child.getAttribute('fill')).toBeDefined();
        expect(child.getAttribute('opacity')).toBeDefined();
        expect(child.getAttribute('r')).toBeDefined();
        expect(child.getAttribute('cx')).toBeDefined();
        expect(child.getAttribute('cy')).toBeDefined();
    }
});
