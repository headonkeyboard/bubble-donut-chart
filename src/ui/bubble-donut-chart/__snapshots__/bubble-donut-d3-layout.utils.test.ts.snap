// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createD3SvgRoot 1`] = `
<div>
  <svg
    height="200"
    width="200"
  />
</div>
`;
